src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -244,65 +244,79 @@ class DecisionTree(BaseModel):
-                                                           cur_node, True)
-                        refit_objs.append(cur_node.left_child)
-                    # If there already is a refit object, add this new index to
-                    # the refit object
-                    elif isinstance(cur_node.left_child, refit_object):
-                        cur_node.left_child.add_idx(idx)
-
-                    cur_node = cur_node.left_child
-                else:
-                    if cur_node.right_child is None:
-                        cur_node.right_child = refit_object(idx, depth,
-                                                            cur_node, False)
-                        refit_objs.append(cur_node.right_child)
-                    elif isinstance(cur_node.right_child, refit_object):
-                        cur_node.right_child.add_idx(idx)
-
-                    cur_node = cur_node.right_child
-                depth += 1
-
-        all_idx = np.arange(0, indices.shape[0], dtype=np.int32)
+        if self.root is not None:
+            refit_objs = []
+            for idx in indices:
+                cur_node = self.root
+                depth = 1
+                while isinstance(cur_node, DecisionNode) :
+                    # Mark cur_node as visited
+                    cur_node.visited = 1
+                    cur_split_idx = cur_node.split_idx
+                    cur_threshold = cur_node.threshold
+
+                    # Check if X should go to the left or right
+                    if X[idx, cur_split_idx] < cur_threshold:
+                        # If the left or right is none, then there previously was a
+                        # leaf node, and we create a new refit object
+                        if cur_node.left_child is None:
+                            cur_node.left_child = refit_object(idx, depth,
+                                                               cur_node, True)
+                            refit_objs.append(cur_node.left_child)
+                        # If there already is a refit object, add this new index to
+                        # the refit object
+                        elif isinstance(cur_node.left_child, refit_object):
+                            cur_node.left_child.add_idx(idx)
+
+                        cur_node = cur_node.left_child
+                    else:
+                        if cur_node.right_child is None:
+                            cur_node.right_child = refit_object(idx, depth,
+                                                                cur_node, False)
+                            refit_objs.append(cur_node.right_child)
+                        elif isinstance(cur_node.right_child, refit_object):
+                            cur_node.right_child.add_idx(idx)
+
+                        cur_node = cur_node.right_child
+                    depth += 1
+
+        all_idx = np.arange(0, X.shape[0], dtype=np.int32)
```

Comment:
changing the all_idx to be all of X, means that you can not pass a indices, which has a smaller size than the X given. This might not be a desired use case, as one might want to pass in the entire X to avoid copying over.

Reply:
Good catch. I updated this now and also fixed this behavior at some other points. I also fixed that the sample weight are taken into account properly in the refit_leaf_nodes now.